sw/device/lib/dif/dif_pinmux.h

```diff
@@ -0,0 +1,502 @@
+ * design.
+ */
+typedef struct dif_pinmux_params {
+  /**
+   * The base address for the a Pin Multiplexer hardware registers.
+   */
+  mmio_region_t base_addr;
+} dif_pinmux_params_t;
+
+/**
+ * Pin Multiplexer Padring pad kinds.
+ *
+ * A pad can be a Multiplexed (MIO) or Dedicated (DIO).
+ */
+typedef enum dif_pinmux_padring_pad_kind {
+  /**
+   * Multiplexed Input/Output pad. Connected to a peripheral input or output
+   * via a multiplexer matrix inside the a Pin Multiplexer.
+   */
+  kDifPinmuxPadringPadKindMio = 0,
+  /**
+   * Dedicated Input/Output pad. Connected directly to a peripheral input
+   * or output, bypassing the multiplexer matrix.
+   */
+  kDifPinmuxPadringPadKindDio,
+} dif_pinmux_padring_pad_kind_t;
+
+/**
+ * Pin Multiplexer functionality locking.
+ *
+ * Represents a type of functionality that can be locked inside the
+ * Pin Multiplexer for a given pad/peripheral input.
+ */
+typedef enum dif_pinmux_lock_type {
+  /**
+   * Lock Padring MIO input select for a given peripheral input. This means that
+   * the peripheral input connection can no longer be changed.
+   */
+  kDifPinmuxLockTypeInsel = 0,
+  /**
+   * Lock peripheral output select for a given Padring MIO output. This means
+   * that the Padring MIO output connection can no longer be changed.
+   */
+  kDifPinmuxLockTypeOutsel,
+  /**
+   * Lock the sleep mode configuration for a given Padring MIO pad.
+   */
+  kDifPinmuxLockTypeMioSleep,
+  /**
+   * Lock the sleep mode configuration for a given Padring DIO pad.
+   */
+  kDifPinmuxLockTypeDioSleep,
+  /**
+   * Lock physical characteristics configuration for a given Padring MIO pad.
+   */
+  kDifPinmuxLockTypeMioPadAttr,
+  /**
+   * Lock physical characteristics configuration for a given Padring DIO pad.
+   */
+  kDifPinmuxLockTypeDioPadAttr,
+} dif_pinmux_lock_type_t;
+
+/**
+ * Pin Multiplexer generic index.
+ *
+ * The meaning of this index changes between APIs, and must be documented in
+ * the corresponding function description.
+ */
+typedef uint32_t dif_pinmux_index_t;
+
+/**
+ * Multiplexed Padring input connection select for a peripheral input.
+ */
+typedef uint32_t dif_pinmux_padring_insel_t;
+
+/**
+ * Multiplexed peripheral output connection select for a Padring output.
+ */
+typedef uint32_t dif_pinmux_peripheral_outsel_t;
+
+/**
+ * Pin multiplexer Padring pad attributes.
+ *
+ * Some attributes have defined meanings, others are implementation defined. Not
+ * all pads will support all attributes. The Pinmux DIF does not check for
+ * unsupported attributes. To check to see if an attribute is supported after
+ * enabling it (via `dif_pinmux_padring_attr_enable` or
+ * `dif_pinmux_padring_attr_write`) read back the attribute field (via
+ * `dif_pinmux_padring_attr_get`). If the attribute isn't supported the
+ * attribute will still be disabled (relevant bit unset in the attribute field).
+ */
+typedef enum dif_pinmux_padring_pad_attr {
```

Comment:
Is the plan here to `|` these together? Maybe the documentation should mention more explicitly that these are bit fields that can be used that way.